### ISSUE TITLE
Multipolygon rendering of outer when inside of an inner

### ISSUE THREAD

**Issue by hungerburg** [gravitystorm/openstreetmap-carto#4416]
A meadow, which is an inner of a forest, contains a woody area, which is mapped as outer of the forest.

### Expected behavior

The woody area should be rendered just like the forest

### Actual behavior

The woody area gets the background colour of the meadow and the icons of the forest

### Links and screenshots illustrating the problem

https://www.openstreetmap.org/way/120475768 - Rendering prompted me to tag the inner, but OSMI told me, that this is wrong. From reading the specification of multipolygon, I believe OSMI is right and OSM-Carto is wrong. The woody area actually is forest, in the meaning of managed for extraction of timber, btw. Also it is densely covered, nowhere thin like the rendering might suggest.

**Comment by imagico** [comment]
I think this is a duplicate of #1796.

The issue description is a bit confusing - mixing real world semantics (meadow, forest) with data model semantics (inner, outer).  Also the modeling of the forest is unnecessarily complex (MP with multiple but topologically nested outer rings - while this is technically valid it is very hard to maintain and various software could have troubles with that).

Anyway - the bottom line is you have a landuse=meadow polygon:

https://www.openstreetmap.org/way/28617016

and a landuse=forest polygon:

https://www.openstreetmap.org/relation/2613608

which overlap and are rendered here exactly as intended:  With the fill color of the meadow above the fill color of the forest (size based drawing order, the forest is larger than the meadow) and the symbols overlay of the forest being drawn above the meadow explicitly and intentionally showing the overlap.

**Comment by imagico** [comment]
Addendum:  Since the suggestion was to render with the forest fill above the meadow fill - that would require abandoning the size based drawing order in the landcover layer (see #4371 - which suggested that in the exact opposite direction) or revisiting #59 - which we most certainly also don't want to do.

**Comment by hungerburg** [comment]
I see, this is treated the same as if the wood had not been cut out from the multipolygon, but just laid there over the meadow. This indeed might be the more common case to find in the wild. The heuristics sure have a reason. So, the way to have it show correctly on the openstreetmap front map would be, make the (inner) meadow a multipolygon too and cut out the wooded part.

The multipolygons in this area are mostly ten years old, people then liked them a lot, perhaps to save on something. They make many edits a chore ;)

**Comment by mboeringa** [comment]
> The multipolygons in this area are mostly ten years old, people then liked them a lot

MPs are still the only sane and unambiguous way to solve these issues. Sorting by size is just a stopgap...

**Comment by imagico** [comment]
@hungerburg - I think you probably have an incorrect understanding of the current mapping here but since you mix real world semantics with data model semantics in your writing i can't tell for sure.  You have two rendering-relevant features here:

https://www.openstreetmap.org/way/28617016
https://www.openstreetmap.org/relation/2613608

Both of these are polygons which overlap at the closed (and untagged - therefore itself not rendered) way:

https://www.openstreetmap.org/way/120475768

**Comment by hungerburg** [comment]
@imagico I refer to this https://wiki.openstreetmap.org/wiki/Relation:multipolygon#Island_within_a_hole - There it says, "where the "island" is the same stuff [as what is outside, so I understand this] it can just be made a hole in the hole." Does the documentation need an update or made more comprehensive? BTW, mboeringa meanwhile chose to change the facts. Will see soon how that turns out.

**Comment by hungerburg** [comment]
@mboeringa When I say chore, I do not think about "pond in a forest" but of tens of km² sized forests, with many outers shared with their neighbouring forests, screes, bare_rocks etc. and even more inners. Openstreetmap rendering of changes in them may sometimes lag for two weeks to.

Looking at your changes, the only thing that comes to mind that makes the outcome now true to the intention, the "inner" meadow now is a multipolygon too, before it was a simple area. Doing a full revert in JOSM, there it looks just the same.

**Comment by hungerburg** [comment]
Extensive panning found me another way of the kind, where an outer of a multipolygon is fully contained within an inner of same mp - https://www.openstreetmap.org/way/137686277 - but is not rendered the same as the other outers - please do not correct! The case is rare, quite often so it seems, mapping has been made to look correct in osm-carto. Maybe though, the documentation cited above is wrong, or I got it wrong, so please leave the case alone, until issue is decided - or documentation is made clear.

**Comment by mboeringa** [comment]
> A "tagless" island inside of an island gets its area type from the multipolygon it is an outer part of

No, it doesn't. There is no "inheritance" of tags between different objects. The fill of the larger object just renders over the area of the smaller forest patch, because no donut hole was created there, so from the perspective of the larger meadow object, there aren't any inner forest patches at all, only meadow! 

The effects you see are just the consequence of a couple of things:

- No use of multipolygons where it technically should or would be highly preferable.
- This particular style rendering polygon color fills and patterns (the tree icons) in two different "layers" or drawing cycles, with the icons in second layer on top of all fills.
- OSM carto using size based sorting (smaller objects rendering on top of larger ones).
- The database, depending on osm2pgsql settings during import of OSM data in the rendering database, potentially breaking up "multipolygons" in individual "polygons", where each multipolygon part becomes an independent record. osm2pgsql can do both: maintain "multipolygons" as-is, or break them apart. Breaking them apart increases the size of tables in terms of number of records, but may reduce the size of individual features to a more manageable level (although most modern hardware should be able to cope with most monster polygons with hundreds of parts in OSM).
- Other particular cartographic choices made in this particular style.

Rather than trying to fully understand why openstreetmap-carto creates a particular rendering, I would recommend you to focus not on rendering, but on what you know is technically right. We do not map for the renderer, but to create the most useful database. That in most cases means creating multipolygons for lakes with islands, or larger meadows with patches of forest in them, and thus not rely on the "size based sorting" and other stuff to "maybe" get it visually right...

**Comment by hungerburg** [comment]
As to the concern, that brought me here: I just wanted the small piece of wood inside of the meadow, an on the ground truth, appear as wood on the map, not as a mixture of wood and meadow. The rendering on the map for me is a means to evaluate the data. I see it as a codification of the prose, that the documentation or specification is.

https://wiki.openstreetmap.org/wiki/Relation:multipolygon#Island_within_a_hole to me suggests, that "way 3" (the island in the pond in the forest) is separate from/outside of "way2" (the pond in the forest), just like "way2" is separate from "way1" (the forest). As I read it, it says, if the island is a forest too, there is no need to make "way2" (the pond) a multipolygon relation that spares out the island.

I now think, OSM-Carto is right, in that it considers the "island" having two competing area types, as a matter of what is in the data. My concern is not about how OSM-Carto deals with such complications. The solution you applied is making this more correct, both from data and how it looks on the map. The documentation needs change. "Cascading" is still necessary, even in the case… What do you think?

**Comment by hungerburg** [comment]
Thinking about how to make the documentation correct, how about, in parentheses and small font perhaps: "Note, that this only says, that way 3 is of the same kind as way 1; it does not make way 3 external to way 2, so their areas will overlap."

**Comment by mboeringa** [comment]
> Thinking about how to make the documentation correct, how about, in parentheses and small font perhaps: "Note, that this only says, that way 3 is of the same kind as way 1; it does not make way 3 external to way 2, so their areas will overlap."

Rather then discussing it here in the style repository, how about adding your comments to the Discussion page of the Wiki page?:
https://wiki.openstreetmap.org/wiki/Talk:Relation:multipolygon
That seems to me to be a better place to discuss it, and possibly get a response from those people who previously edited the page.

By the way, you're misinterpreting the "Island within a hole" example: way 2 is nothing by itself in this example, _**as it has no tags**_! As such, it will not be interpreted as a separate area / polygon object by any OSM data importer, so there is no _"overlap"_.

It is just a constituent ring or the multipolygon with **"relation id=1"**, _which is the only object defined here in this example_, as can be clearly seen from the XML:

```
<relation id="1">
  <tag k="type" v="multipolygon" />
  <member type="way" id="1" role="outer" />
  <member type="way" id="2" role="inner" />
  <member type="way" id="3" role="outer" />
</relation>
```
